pkg/object/mqttproxy/broker.go

```diff
@@ -0,0 +1,336 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package mqttproxy
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+
+	"github.com/eclipse/paho.mqtt.golang/packets"
+	"github.com/megaease/easegress/pkg/api"
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/object/function/storage"
+)
+
+type (
+	// Broker is MQTT server, will manage client, topic, session, etc.
+	Broker struct {
+		sync.RWMutex
+		egName string
+		name   string
+		spec   *Spec
+
+		listener net.Listener
+		backend  BackendMQ
+		clients  map[string]*Client
+		auth     map[string]string
+		tlsCfg   *tls.Config
+
+		sessMgr   *SessionManager
+		topicMgr  *TopicManager
+		memberURL func(string, string) ([]string, error)
+
+		// done is the channel for shutdowning this proxy.
+		done chan struct{}
+	}
+
+	// HTTPJsonData is json data received from http endpoint used to send back to clients
+	HTTPJsonData struct {
+		Topic       string `json:"topic"`
+		Qos         int    `json:"qos"`
+		Payload     string `json:"payload"`
+		Base64      bool   `json:"base64"`
+		Distributed bool   `json:"distributed"`
+	}
+)
+
+func newBroker(spec *Spec, store storage.Storage, memberURL func(string, string) ([]string, error)) *Broker {
+	broker := &Broker{
+		egName:    spec.EGName,
+		name:      spec.Name,
+		spec:      spec,
+		backend:   newBackendMQ(spec),
+		clients:   make(map[string]*Client),
+		auth:      make(map[string]string),
+		memberURL: memberURL,
+		done:      make(chan struct{}),
+	}
+	broker.topicMgr = newTopicManager()
+	broker.sessMgr = newSessionManager(broker, store)
+	for _, a := range spec.Auth {
+		broker.auth[a.Username] = a.B64Passwd
+	}
+	err := broker.setListener()
+	if err != nil {
+		logger.Errorf("mqtt.newBroker: broker set listener failed, err:%v", err)
+		return nil
+	}
+
+	go broker.run()
+	return broker
+}
+
+func (b *Broker) setListener() error {
+	var l net.Listener
+	var err error
+	var cfg *tls.Config
+	addr := fmt.Sprintf(":%d", b.spec.Port)
+	if b.spec.UseTLS {
+		cfg, err = b.spec.tlsConfig()
+		if err != nil {
+			return fmt.Errorf("invalid tls config for mqtt proxy, err:%v", err)
+		}
+		l, err = tls.Listen("tcp", addr, cfg)
+		if err != nil {
+			return fmt.Errorf("gen mqtt tls tcp listener failed, addr:%s, err:%v", addr, err)
+		}
+	} else {
+		l, err = net.Listen("tcp", addr)
+		if err != nil {
+			return fmt.Errorf("gen mqtt tcp listener failed, addr:%s, err:%v", addr, err)
+		}
+	}
+	b.tlsCfg = cfg
+	b.listener = l
+	return err
+}
+
+func (b *Broker) run() {
+	for {
+		conn, err := b.listener.Accept()
+		if err != nil {
+			select {
+			case <-b.done:
+				return
+			default:
+				logger.Errorf("mqtt.run: net listener accept failed, err:%s", err)
+			}
+		} else {
+			go b.handleConn(conn)
+		}
+	}
+}
```

Comment:
I'm not a Go expert, just one comment about the code style:
It seems that the whole project use tab instead of 4 spaces as the indent. Could anyone explain the reason?

Github uses 8 spaces to render a tab. This block of the code with tab indent seems really ugly.

Reply:
The reason is in my VSCode, tab is set to 4 spaces. I will update that.

Reply:
As said in Effective Go https://golang.org/doc/effective_go#formatting
"Indentation
We use tabs for indentation and gofmt emits them by default. Use spaces only if you must."

In Easegress, we use tab to indent, you can use your IDE to config tab with 4 spaces if you like.